![header](https://magazin.ebay-kleinanzeigen.de/app/uploads/sites/8/2018/02/ebay-kleinanzeigen_logo_rgb.jpg)

# Context:

In this project, we will analyze an dataset from **eBay Kleinanzeigen**. The dataset can be found on [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data) but was cleaned in advance for the purpose of the analysis. This data set was scraped in **2016** (this analysis was made on August 2020). If a new data set were to be prepared this year, data could be entirely different. 

The data set is composed of the following columns: 

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

# Purpose: 

The very purpose of this project is to display my skills in data cleaning with a quick and efficient analysis of the market. 

In [94]:
## import of the necessary packages 

import numpy as np 
import pandas as pd 

In [95]:
## let's read the data set with pandas 

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [96]:
autos.head() #.head() will help us understand the data composition by dispalying the first 5 rows (headers inclucded)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [97]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

The data is composed of 20 columns. Some of those columns contain null values but it does not seem that it exceeds 20%, which will make relatively easy to clean. 

One issue is that the names are written in CamelCaps. We will modify the shape of the name into [snakecase](https://en.wikipedia.org/wiki/Snake_case) (DateCrawled -> date_crawled) to make our analysis more comfortable. 

In [98]:
autos.columns #were we are going to display the column names to turn them into snakecase

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [99]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

**The names are now changed which makes it easier and clearer for us to run an anlysis.**

In [100]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Most data columns looks quite alright. However, we can see that the column "*price*" is considered as a string as the symbol $ appears before the figures. Also, the numbers are separated by a coma. 

We have the same issue with the "*odometer*" column where we can a "km" before the figures. The numbers are also in a incorrect format, being separated by a coma. 

In [101]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [102]:
autos["odometer"] = (autos["odometer"]
                          .str.replace("km","")
                          .str.replace(",","")
                          .astype(int)
                          )

autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

**Both columns are now in the correct shape (format *int*)**. We can now focus a little bit more and those two colums and analyze the price and odometer data of the cars that are being auctioned on the Ebay German website. 

# Price and mileage 

In [103]:
print("Unique values:", autos['price'].unique().shape)

Unique values: (2357,)


In [104]:
print(autos['price'].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [105]:
print(autos['price'].value_counts().sort_index(ascending=True))

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

The "price" column has a few strange values. 1421 cars have a price of 0 EUR. Since Ebay is an auction website, it is not so surprising. However, cars at a price of 0 are likely to have been posted quite recently. 

Also, There is an important gap between prices. 

![description](https://loicchamplong.com/wp-content/uploads/2020/08/price_gap.png)

On the image we can see that prices jump from 350,000EUR to 999,990 and reaching 99 million euros. Even for an auction website with special car collection, this gap and those figures look quite odd. 

For the purpose of the analysis, we will clean this column by keeping cars in a price range of 1 to 350,000 EUR. 

In [106]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [107]:
print("Unique values:", autos['odometer_km'].unique().shape)

Unique values: (13,)


In [108]:
autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [109]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
20000       762
10000       253
5000        836
Name: odometer_km, dtype: int64

**There is nothing peculiar to be noted with the "odometer_km" column. We will keep the values as they are**. 

# Date Columns

There are columns with date values that we are not going to clean and analyze. 

In [110]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [111]:
(autos['date_crawled'].str[:10]
                      .value_counts(normalize=True, dropna=False)
                      .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [112]:
(autos['ad_created'].str[:10]
                      .value_counts(normalize=True, dropna=False)
                      .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

In [113]:
(autos['last_seen'].str[:10]
                      .value_counts(normalize=True, dropna=False)
                      .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

It appears that March 2016 was a month with lots of traffic onto the Ebay website. Despite from that, dates are mostly uniformed and there is nothing perculiar about. 

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [114]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [115]:
autos = autos[autos["registration_year"].between(1960,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067676
2005    0.062958
1999    0.062122
2004    0.057962
2003    0.057876
2006    0.057254
2001    0.056525
2002    0.053309
1998    0.050671
2007    0.048827
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

**Here, we have modified the date by selecting the cars registered in 1960 to 2016**. It makes much more sense. In the end it turns out that the oldest car was from 1998 and the most recent 2007. Most of the cars are from the years 1999 to 2000 

# The most expensive brand sold 

Now that we are a better overview of the date and price of the cars, we can analyze the brands that were the most expensive (hence most popular since it is an auction website) on the auction website. 

In [116]:
brand_counts = autos["brand"].value_counts(normalize=True, ascending=False)
brand_counts

volkswagen        0.211434
bmw               0.110134
opel              0.107604
mercedes_benz     0.096389
audi              0.086654
ford              0.069842
renault           0.047176
peugeot           0.029871
fiat              0.025668
seat              0.018270
skoda             0.016404
nissan            0.015289
mazda             0.015203
smart             0.014174
citroen           0.014003
toyota            0.012716
hyundai           0.010036
sonstige_autos    0.009457
volvo             0.009156
mini              0.008770
mitsubishi        0.008234
honda             0.007848
kia               0.007076
alfa_romeo        0.006626
porsche           0.006111
suzuki            0.005940
chevrolet         0.005683
chrysler          0.003517
dacia             0.002638
daihatsu          0.002509
jeep              0.002273
subaru            0.002144
land_rover        0.002080
saab              0.001651
jaguar            0.001565
daewoo            0.001501
trabant           0.001372
r

It appears that German brands are the most popular cars to be bought on the website. We are going to analyze further in order to determine the most popular brands by selecting the top 20% of car brands sold. 

In [117]:
most_popular_brand = brand_counts[brand_counts > .02].index
most_popular_brand

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat'],
      dtype='object')

*'volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
'peugeot'* and *'fiat'* are the most expensoive brands. 

In [118]:
brand_mean_prices = {}

for brand in most_popular_brand:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices 

{'audi': 9336,
 'bmw': 8332,
 'fiat': 2813,
 'ford': 3713,
 'mercedes_benz': 8565,
 'opel': 2974,
 'peugeot': 3094,
 'renault': 2475,
 'volkswagen': 5398}

In [119]:
bmp_series = pd.Series(brand_mean_prices) #we construct a pandas serie in order to create a dataframe
pd.DataFrame(bmp_series, columns=["mean_price"]).sort_values('mean_price', ascending=False) #we create the dataframe of "Mean Price"

mean_price
audi                 9336
mercedes_benz        8565
bmw                  8332
volkswagen           5398
ford                 3713
peugeot              3094
opel                 2974
fiat                 2813
renault              2475

It appears that German brands are indeed the most popular being on the top 3. There are also a american brand (Ford) and two french manufacturers (Peugeot and Renault). Fiat (Italy) is also quite popular. 

# Mileage 

Now that we know which cars are the most popular and the most expensive ones, does the mileage have an impact on the prices? 

In [120]:
brand_mean_mileage = {}

for brand in most_popular_brand:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [121]:
brand_mileage = pd.DataFrame(mean_mileage, columns=['brand_mileage'])
brand_mileage

brand_mileage
bmw                   132597
mercedes_benz         130919
opel                  129342
audi                  129157
volkswagen            128707
renault               128127
peugeot               127153
ford                  124399
fiat                  117121

In [122]:
brand_mileage["mean_price"] = mean_prices
brand_mileage

brand_mileage  mean_price
bmw                   132597        8332
mercedes_benz         130919        8565
opel                  129342        2974
audi                  129157        9336
volkswagen            128707        5398
renault               128127        2475
peugeot               127153        3094
ford                  124399        3713
fiat                  117121        2813

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.